File: javascript/clock/clock.js
class Clock {
    constructor(hours, minutes) {
        this.hours = hours;
        this.minutes = minutes;
        this.adjust();
    }

    adjust() {
        while (this.minutes < 0) {
            this.minutes += 60;
            this.hours--;
        }
        while (this.minutes >= 60) {
            this.minutes -= 60;
            this.hours++;
        }

        while (this.hours < 0) {
            this.hours += 24;
        }
        while (this.hours >= 24) {
            this.hours -= 24;
        }
    }

    toString() {
        return `${this.hours < 10 ? '0' : ''}${this.hours}:${this.minutes < 10 ? '0' : ''}${this.minutes}`;
    }

    plus(minutes) {
        this.minutes += minutes;
        this.adjust();
        return this;
    }
    minus(minutes) {
        this.minutes -= minutes;
        this.adjust();
        return this;
    }

    equals(other) {
        if (!(other instanceof Clock)) {
            return false;
        }
        return this.hours === other.hours && this.minutes === other.minutes;
    }
}

export function at(hours = 0, minutes = 0) {
    return new Clock(hours, minutes);
}

File: java/clock/src/main/java/Clock.java
import java.util.Objects;

class Clock {
    public static final int MINUTES_PER_HOUR = 60;
    public static final int HOURS_PER_DAY = 24;
    public static final int MINUTES_PER_DAY = MINUTES_PER_HOUR * HOURS_PER_DAY;

    private int t;

    Clock(int h, int m) {
        t = (h * MINUTES_PER_HOUR) + m;
        adjust();
    }

    void add(int m) {
        t += m;
        adjust();
    }

    @Override
    public String toString() {
        return String.format("%02d:%02d", t / MINUTES_PER_HOUR, t % MINUTES_PER_HOUR);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Clock clock = (Clock) o;
        return t == clock.t;
    }

    @Override
    public int hashCode() {
        return Objects.hash(t);
    }

    private void adjust() {
        while (t < 0) {
            t += MINUTES_PER_DAY;
        }
        t = t % MINUTES_PER_DAY;
    }
}